### ISSUE TITLE
IMAP4 Body parts fetch error

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10895]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#1307|
|Type|defect|
|Created|2005-11-02 21:13:16Z|

Attachments:

* [test_imap.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/40b/40b4efaaa821a3b5b1282b8c60e1b82c3b05baae/b3bb7c9dad50d5d05881a2e5f3a3369a6fdf20e5.diff) (3541 bytes) - added by therve on 2005-11-03 19:03:50Z - 
* [imap4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/40b/40b4efaaa821a3b5b1282b8c60e1b82c3b05baae/859a30cdb399d2559e889de25bb96ab68071ad92.diff) (3035 bytes) - added by therve on 2005-11-03 19:05:23Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1307 1307
type__defect defect
reporter__therve therve
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1130965996000000 1130965996000000
changetime__1142126817000000 1142126817000000
version__ 
owner__ 
cc__exarkun cc__jojo cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

```
#!html
<pre>
A multiple part fetch with IMAP4 client failed with this error: 
2005/11/02 15:00 CET [SimpleIMAP4Client,client] Traceback (most recent call last):
          File "/usr/lib/python2.3/site-packages/twisted/protocols/basic.py",
line 247, in setLineMode
            return self.dataReceived(extra)
          File "/usr/lib/python2.3/site-packages/twisted/protocols/basic.py",
line 221, in dataReceived
            why = self.lineReceived(line)
          File "/usr/lib/python2.3/site-packages/twisted/mail/imap4.py", line
2070, in lineReceived
            self._regularDispatch(line)
          File "/usr/lib/python2.3/site-packages/twisted/mail/imap4.py", line
2100, in _regularDispatch
            self.dispatchCommand(tag, rest)
        --- &lt;exception caught here> ---
          File "/usr/lib/python2.3/site-packages/twisted/mail/imap4.py", line
2129, in dispatchCommand
            f(tag, rest)
          File "/usr/lib/python2.3/site-packages/twisted/mail/imap4.py", line
2156, in response_UNAUTH
            self._defaultHandler(tag, rest)
          File "/usr/lib/python2.3/site-packages/twisted/mail/imap4.py", line
2177, in _defaultHandler
            raise IllegalServerResponse(tag + ' ' + rest)
        twisted.mail.imap4.IllegalServerResponse: &lt;data>

The fetch used is in the form : "FETCH &lt;NUM> (BODY[i] BODY[j])". It's handled
correctly by standard imaplib. I join a "naive" patch that corrects my needs.

-- 
Thomas
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
I'd like to fix this, but I lack the time to do it properly.  I'm wary of
applying a patch to this part of the code since it is not heavily tested, it is
complex, and I am not even sure what this response looks like.

Could you attach the server's response, and possibly a unit test for this behavior?

Otherwise this will probably be put off until I have time to do a serious
attempt at redoing the parser.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

```
#!html
<pre>
OK, I can first show you a telnet session. I connect to my local dovecot server,
 fetched the bodystructure of mail message that shows one text part and its
alternative in html. I retrieve those using the command that fails with Twisted:
 
$ telnet localhost 143
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
* OK Dovecot ready.
* LOGIN username password
* OK Logged in.
* SELECT INBOX
* FLAGS (\Answered \Flagged \Deleted \Seen \Draft unknown-0 unknown-1 unknown-2
unknown-4 $MDNSent Junk)
* OK [PERMANENTFLAGS (\Answered \Flagged \Deleted \Seen \Draft unknown-0
unknown-1 unknown-2 unknown-4 $MDNSent Junk \*)] Flags permitted.
* 401 EXISTS
* 0 RECENT
* OK [UIDVALIDITY 1106242061] UIDs valid
* OK [UIDNEXT 9794] Predicted next UID
* OK [READ-WRITE] Select completed.
* FETCH 396 BODYSTRUCTURE
* 396 FETCH (BODYSTRUCTURE (("text" "plain" ("charset" "ISO-8859-1") NIL NIL
"7bit" 2144 62 NIL NIL NIL)("text" "html" ("charset" "ISO-8859-1") NIL NIL
"7bit" 12649 155 NIL NIL NIL) "alternative" ("boundary"
"----=_NextPart_84815C5ABAF209EF376268C8") NIL NIL))
* OK Fetch completed.
* FETCH 396 (BODY[1] BODY[2])
* 396 FETCH (BODY[1] {2144}
&lt;blabla plain text of 2144 bytes> BODY[2] {12649}
&lt;blabla text/html of 12649 bytes>)
* OK Fetch completed.
* LOGOUT
* BYE Logging out
* OK Logout completed.
Connection closed by foreign host.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

```
#!html
<pre>
Joined a quick patch to test_imap.py that approximatively shows what I think
could be the good behaviour. It doesn't run even with my modifications because
"fetch (body[i] body[j])" doesn't work with IMAP4Server.

Note: fetchSpecific doesn't have a unittest. Is there a problem with it or it
just a miss ?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve commented|
|-|-|

```
#!html
<pre>
And a new patch that feels better with trial.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16260]) Properly parse and serve the response to FETCH X (BODY[N] BODY[M])

Mostly this is a patch from therve, but with modifications to actually have the server spew the
body of the correct section of the message.

Fixes [#10895](https://github.com/twisted/twisted/issues/10895)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
